### DISCUSSION TITLE
Shared Service - Vanilla Hazelcast

### DISCUSSION THREAD

**Discussion by charliecatney** [vladimir-bukhtoyarov/bucket4j#218] [Q&A] [answered]
Hi,

I work in an environment where we have a managed Hazelcast cluster, consequently it is not straightforward to add Bucket4J jars to the cluster nodes class path.

Does bucket4j offer a distributed model with all serialisation etc happening on the client side, or using Hazelcast built in serializers? 

Thanks

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @charliecatney 

Currently, there is no such feature. But I can add it. However, you need to get a deep understanding that client-side implementations based on CAS or locking will show less performance than current implementation because bucket4j-hazelcast uses EntryProcessors, see https://hazelcast.com/blog/an-easy-performance-improvement-with-entryprocessor/

If you agree with performance degradation, I will add such a feature(most likely at weekend).

**Reply by charliecatney** [reply]
Hi @vladimir-bukhtoyarov,

That would be great. I appreciate the principals and understand the EntityProcessor model gives significant performance gains.

It would be interesting to do some performance benchmarking of the 2 approaches, to understand the material impact. In the Bucket4j Token Bucket implementation, does every request for a permit make a call to the backing store?

Thanks

**Reply by vladimir-bukhtoyarov** [reply]
Yes. One permit check leads to one interaction with store until one of [optimizations](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization/Optimizations.java) is applied [during bucket construction](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L61).

**Reply by charliecatney** [reply]
Thank You, @vladimir-bukhtoyarov, let me know once you have had a chance to look at the client side implementation, and I really appreciate the quick response to my query.

**Reply by vladimir-bukhtoyarov** [reply]
@charliecatney 

Desired functionality has been released with version ```7.1.0```

See documentation https://bucket4j.com/7.1.0/toc.html#support-for-externally-managed-hazelcast-without-classpath-access

**Reply by charliecatney** [reply]
Awesome, thanks @vladimir-bukhtoyarov. I'll integrate with this and let you know if I encounter any issues.
